File: scripts/createUser.mjs
import fs from 'fs';
import path from 'path';
import url from 'url';
import yargs from 'yargs';
import { hideBin } from 'yargs/helpers';
import admin from 'firebase-admin';

// Usage:
// node scripts/createUser.mjs --email x@y.com --password 'secret' --fullName 'Name' --reviewLink 'https://...'
//   [--username 'name'] [--senderName 'Your Practice'] [--no-sendEmail]
// Requires service account JSON at ./serviceAccountKey.json

const __dirname = path.dirname(url.fileURLToPath(import.meta.url));
const projectRoot = path.resolve(__dirname, '..');
const serviceKeyPath = path.join(projectRoot, 'serviceAccountKey.json');

if (!fs.existsSync(serviceKeyPath)) {
  console.error('Missing serviceAccountKey.json at project root. Download from Firebase Console > Project settings > Service accounts.');
  process.exit(1);
}

const serviceAccount = JSON.parse(fs.readFileSync(serviceKeyPath, 'utf8'));

if (!admin.apps.length) {
  admin.initializeApp({
    credential: admin.credential.cert(serviceAccount)
  });
}

const argv = yargs(hideBin(process.argv))
  .option('email', { type: 'string', demandOption: true })
  .option('password', { type: 'string', demandOption: true })
  .option('fullName', { type: 'string', demandOption: true })
  .option('reviewLink', { type: 'string', demandOption: true })
  .option('username', { type: 'string', describe: 'Optional username to assign (lowercased, a-z0-9._-)' })
  .option('senderName', { type: 'string', default: 'QuickReviews', describe: 'Friendly From name for welcome email' })
  .option('sendEmail', { type: 'boolean', default: true, describe: 'Send welcome email with reset link' })
  .strict()
  .argv;

async function main() {
  const { email, password, fullName, reviewLink, username, senderName, sendEmail } = argv;

  // Create Auth user (idempotent on email)
  let userRecord;
  try {
    userRecord = await admin.auth().getUserByEmail(email);
    console.log(`User exists: ${userRecord.uid}`);
  } catch (err) {
    if (err.code === 'auth/user-not-found') {
      userRecord = await admin.auth().createUser({ email, password, emailVerified: true });
      console.log(`Created user: ${userRecord.uid}`);
    } else {
      throw err;
    }
  }

  // Write Firestore profile (doc can be ID = uid or existing email doc)
  const db = admin.firestore();

  // Try to find existing profile by email; otherwise create new doc with uid
  const qs = await db.collection('users').where('email', '==', email).limit(1).get();
  const docRef = qs.empty ? db.collection('users').doc(userRecord.uid) : qs.docs[0].ref;

  const profileUpdates = {
    email,
    fullName,
    reviewLink,
    needsPasswordChange: true
  };

  // Assign username if provided; create mapping document
  let finalUsername = null;
  if (username) {
    const desired = String(username).toLowerCase();
    if (!/^[a-z0-9._-]{3,30}$/.test(desired)) {
      throw new Error('Invalid --username. Use 3–30 of a-z, 0-9, dot, underscore, hyphen.');
    }
    let candidate = desired;
    // Ensure uniqueness by appending digits if needed
    for (let i = 0; i < 50; i++) {
      const doc = await db.collection('usernames').doc(candidate).get();
      if (!doc.exists) { finalUsername = candidate; break; }
      candidate = `${desired}${Math.floor(1000 + Math.random()*9000)}`;
    }
    if (!finalUsername) throw new Error('Could not assign a unique username after multiple attempts.');
    await db.collection('usernames').doc(finalUsername).set({
      uid: userRecord.uid,
      email,
      createdAt: admin.firestore.FieldValue.serverTimestamp(),
    });
    profileUpdates.username = finalUsername;
    profileUpdates.usernameLower = finalUsername;
  }

  await docRef.set(profileUpdates, { merge: true });

  console.log('Profile upserted at:', docRef.path);

  // Generate a password reset link (safer than sharing raw temp password)
  let resetLink = null;
  try {
    resetLink = await admin.auth().generatePasswordResetLink(email);
    console.log('Password reset link (optional to send to user):');
    console.log(resetLink);
  } catch (e) {
    console.warn('Could not generate password reset link:', e.message);
  }

  // Optionally send welcome/invite email via Firestore mail collection
  if (sendEmail && resetLink) {
    const from = `${senderName} <feedback@ezreviews.app>`;
    const html = `
      <div style="font-family:Inter,system-ui,Segoe UI,Arial,sans-serif;">
        <p>Hi ${fullName.split(' ')[0] || ''},</p>
        <p>Welcome to QuickReviews. Click the link below to set your password and get started:</p>
        <p><a href="${resetLink}">Set your password</a></p>
        ${finalUsername ? `<p>Your username: <b>${finalUsername}</b></p>` : ''}
        <p>Thanks!</p>
      </div>
    `;
    await db.collection('mail').add({
      to: email,
      from,
      message: { subject: 'You are invited to QuickReviews', html }
    });
    console.log('Welcome email enqueued.');
  }
}

main().catch(err => {
  console.error(err);
  process.exit(1);
});

File: scripts/create_user.py
import argparse
import os
import sys
import json
import string
import secrets

import firebase_admin
from firebase_admin import credentials, auth, firestore


def generate_temp_password(length: int = 8) -> str:
    # Use only letters and numbers, no special characters
    alphabet = string.ascii_letters + string.digits
    # Ensure at least one letter and one number
    password = []
    password.append(secrets.choice(string.ascii_lowercase))  # One lowercase letter
    password.append(secrets.choice(string.digits))           # One number
    # Fill the rest randomly
    for _ in range(length - 2):
        password.append(secrets.choice(alphabet))
    # Shuffle the password
    password_list = list(password)
    secrets.SystemRandom().shuffle(password_list)
    return ''.join(password_list)


def main() -> None:
    parser = argparse.ArgumentParser(description="Create or update a Firebase Auth user and Firestore profile.")
    parser.add_argument("--email", required=True, help="User email")
    parser.add_argument("--full-name", dest="full_name", default="", help="Full name to store in profile")
    parser.add_argument("--review-link", dest="review_link", default="", help="Review link to store in profile")
    parser.add_argument("--temp-password", dest="temp_password", default=None, help="Optional temporary password to set (random if omitted)")
    parser.add_argument("--service-account", dest="service_account", default=os.path.join(os.path.dirname(os.path.dirname(__file__)), "serviceAccountKey.json"), help="Path to service account JSON (default: project root serviceAccountKey.json)")
    args = parser.parse_args()

    if not os.path.exists(args.service_account):
        print(f"Missing service account JSON at: {args.service_account}", file=sys.stderr)
        sys.exit(1)

    cred = credentials.Certificate(args.service_account)
    if not firebase_admin._apps:
        firebase_admin.initialize_app(cred)

    email = args.email.strip().lower()
    temp_password = args.temp_password or generate_temp_password()
    full_name = args.full_name.strip()
    review_link = args.review_link.strip()

    created = False
    try:
        user = auth.get_user_by_email(email)
        # Reset password to the temporary one
        user = auth.update_user(user.uid, password=temp_password)
        print(f"User exists. Password reset. uid={user.uid}")
    except auth.UserNotFoundError:
        user = auth.create_user(email=email, password=temp_password, email_verified=True)
        created = True
        print(f"Created user. uid={user.uid}")

    db = firestore.client()
    qs = db.collection("users").where("email", "==", email).limit(1).get()
    doc_ref = qs[0].reference if qs else db.collection("users").document(user.uid)

    profile_updates = {"email": email, "needsPasswordChange": True}
    if full_name:
        profile_updates["fullName"] = full_name
    if review_link:
        profile_updates["reviewLink"] = review_link

    doc_ref.set(profile_updates, merge=True)
    print(f"Profile upserted at: {doc_ref.path}")

    # Provide a password reset link admins can send to the user
    try:
        reset_link = auth.generate_password_reset_link(email)
        print("Password reset link (optional to send to user):")
        print(reset_link)
    except Exception as e:
        print(f"Could not generate password reset link: {e}")

    print("\nTemporary password (share securely with the user):")
    print(temp_password)
    print("\nOn first login, user will be prompted to change their password and then complete profile.")


if __name__ == "__main__":
    main()


